WordPress/WordPressStatsWidgets/Helpers/WidgetDataReader.swift

```diff
@@ -0,0 +1,75 @@
+import Foundation
+
+protocol WidgetDataCacheReader {
+    func widgetData<T: HomeWidgetData>(for siteID: String) -> T?
+}
+
+final class WidgetDataReader<T: HomeWidgetData> {
+    let userDefaults: UserDefaults?
+    let cacheReader: WidgetDataCacheReader
+
+    init(_ userDefaults: UserDefaults? = UserDefaults(suiteName: WPAppGroupName),
+         _ cacheReader: any WidgetDataCacheReader = HomeWidgetDataFileReader()
+    ) {
+        self.userDefaults = userDefaults
+        self.cacheReader = cacheReader
+    }
+
+    /// Returns cached widget data based on the selected site when editing widget and the default site.
+    /// Configuration.site is nil until IntentHandler is initialized.
+    /// Configuration.site can have old value after logging in with a different account. No way to reset configuration when the user logs out.
+    /// Using defaultSiteID if both of these cases.
+    /// - Parameters:
+    ///   - configuration: Configuration of the Widget Site Selection Intent
+    ///   - defaultSiteID: ID of the default site in the account
+    /// - Returns: Widget data
+    func widgetData(for configuration: SelectSiteIntent, defaultSiteID: Int?) -> T? {
+
+        /// If configuration.site.identifier has value but there's no widgetData, it means that this identifier comes from previously logged in account
+        if let selectedSite = configuration.site?.identifier,
+           let widgetData: T = cacheReader.widgetData(for: selectedSite) {
+            return widgetData
+        } else if let defaultSiteID = defaultSiteID {
+            return cacheReader.widgetData(for: String(defaultSiteID))
+        } else {
+            return nil
+        }
+    }
+
+    func widgetData(
+        for configuration: SelectSiteIntent,
+        defaultSiteID: Int?,
+        isJetpack: Bool,
+        onDisabled: (() -> Void)? = nil,
```

Comment:
Instead of many callbacks, what would you think about the `Result` type? Or a custom `enum` with many cases? I think it would correctly indicate 2 things:

- The method is synchronous 
- The method only has one return type at the time

Maybe it would also help with testability? Do you think this is a change worth making?

Reply:
Yup, it makes sense, I also feel multiple callbacks can't guarantee one return, so that's why I write the unit test,
I agree with `Result` is better, I will submit an update tomorrow 👍